magefiles/configure.go

```diff
@@ -4,70 +4,229 @@
+	PackageType string `json:"packagetype"`
+}
+
+type PyPIPackage struct {
+	Releases map[string][]Release `json:"releases"`
+}
+
+var errNoReleasesFound = errors.New("no mlflow sdist releases found")
+
+func getMlflowArtifact(ctx context.Context, version string) (string, error) {
+	url := "https://pypi.org/pypi/mlflow/json"
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %w", err)
+	}
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to perform request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var pypiData PyPIPackage
+
+	err = json.Unmarshal(body, &pypiData)
+	if err != nil {
+		return "", err
+	}
+
+	if releases, found := pypiData.Releases[version]; found && len(releases) > 0 {
+		for _, r := range releases {
+			if r.PackageType == "sdist" {
+				return r.URL, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("version %s: %w", version, errNoReleasesFound)
+}
+
+func downloadTarFile(ctx context.Context, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	tempDir := os.TempDir()
+
+	tempFile, err := os.CreateTemp(tempDir, "*.tar.gz")
+	if err != nil {
+		return "", err
+	}
+
+	defer tempFile.Close()
+
+	_, err = io.Copy(tempFile, resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return tempFile.Name(), nil
+}
+
+func extractedTar(ctx context.Context, tarball string) (string, error) {
+	data, err := os.ReadFile(tarball)
+	if err != nil {
+		return "", err
+	}
+
+	buffer := bytes.NewBuffer(data)
+	destination := strings.TrimSuffix(tarball, ".tar.gz")
+
+	err = extract.Gz(ctx, buffer, destination, nil)
+	if err != nil {
+		return "", err
+	}
+
+	return destination, err
+}
+
+// copyFile copies a file from src to dst, preserving file permissions.
+func copyFile(srcFile, dstFile string) error {
+	// Open the source file
+	src, err := os.Open(srcFile)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	// Create the destination file (create directory if needed)
+	dst, err := os.Create(dstFile)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	// Copy file content
+	if _, err := io.Copy(dst, src); err != nil {
+		return err
+	}
+
+	// Set the same file permissions
+	info, err := os.Stat(srcFile)
+	if err != nil {
+		return err
+	}
 
-	return sh.RunV("pip", allArgs...)
+	return os.Chmod(dstFile, info.Mode())
 }
 
-func tar(args ...string) error {
-	return sh.RunV("tar", args...)
+// CopyDirectory recursively copies a source directory to a destination directory.
+func copyDirectory(srcDir, destDir string) error {
```

Comment:
Surely there's a library for this already?

Reply:
@dsuhinin any suggestions?